File: backend/ms-account/src/main/java/com/bantads/service/AccountHistoryService.java
package com.bantads.service;

import com.bantads.dto.read.AccountHistoryDTO;
import com.bantads.dto.read.ExtractDTO;
import com.bantads.entity.event.Event;
import com.bantads.entity.event.EventType;
import com.bantads.entity.read.AccountHistory;
import com.bantads.entity.read.TransactionType;
import com.bantads.repository.read.AccountHistoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.math.BigDecimal;
import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;

@Service
public class AccountHistoryService {

    @Autowired
    private AccountHistoryRepository accountHistoryRepository;

    public AccountHistory createAcountHistory(AccountHistoryDTO ac) {
        AccountHistory accountHistory = AccountHistory.builder()
            .accountNumber(ac.getAccountNumber())
            .amount(ac.getAmount())
            .createdAt(LocalDateTime.now())
            .destinationClientCPF(ac.getDestinationClientCPF())
            .destinationClientName(ac.getDestinationClientName())
            .managerCPF(ac.getManagerCPF())
            .managerName(ac.getManagerName())
            .originClientCPF(ac.getOriginClientCPF())
            .originClientName(ac.getOriginClientName())
            .type(ac.getType())
            .build();

        return accountHistoryRepository.save(accountHistory);
    }

    private LocalDateTime parseDateTime(String dateStr, boolean isStart) {
        if (isStringNull(dateStr)) {
            return isStart
                ? LocalDate.now().minusDays(30).atStartOfDay()
                : LocalDate.now().atTime(LocalTime.MAX);
        }

        String trimmed = dateStr.trim();

        if (trimmed.length() == 10) {
            LocalDate localDate = LocalDate.parse(trimmed);
            return isStart ? localDate.atStartOfDay() : localDate.atTime(LocalTime.MAX);
        }

        if (trimmed.contains(" ") && !trimmed.contains("T")) {
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
            return LocalDateTime.parse(trimmed, formatter);
        }

        return LocalDateTime.parse(trimmed);
    }

    public ExtractDTO getExtract(String accountNumber, String start, String end, String userCPF) {
        LocalDateTime startDate = parseDateTime(start, true);
        LocalDateTime endDate = parseDateTime(end, false);

        this.checkDate(startDate, endDate);

        ExtractDTO extract = new ExtractDTO();
        BigDecimal balance = this.getInitialBalance(accountNumber, startDate, userCPF);
        extract.setSaldoAbertura(balance.toPlainString());

        List<AccountHistory> history = this.accountHistoryRepository
                .findAllByAccountNumberAndCreatedAtBetween(accountNumber, startDate, endDate);

        extract.setMovimentacoes(history);
        return extract;
    }

    public BigDecimal getInitialBalance(String accountNumber, LocalDateTime start, String userCPF) {
        List<AccountHistory> history = this.accountHistoryRepository.findAllByAccountNumberAndCreatedAtLessThan(accountNumber, start);

        Iterator<AccountHistory> historyIterator = history.iterator();

        BigDecimal balance = new BigDecimal(0);

        while(historyIterator.hasNext()) {
            AccountHistory ac = historyIterator.next();
            TransactionType type = ac.getType();
            BigDecimal value = ac.getAmount();

            if(type == TransactionType.DEPOSIT) {
                balance = balance.add(value);
            }

            if(type == TransactionType.WITHDRAW) {
                balance = balance.subtract(value);
            }

            if(type == TransactionType.TRANSFER && Objects.equals(ac.getDestinationClientCPF(), userCPF)) {
                balance = balance.add(value);
            }

            if(type == TransactionType.TRANSFER && Objects.equals(ac.getOriginClientCPF(), userCPF)) {
                balance = balance.subtract(value);
            }
        }

        return balance;
    }

    public void checkDate(LocalDateTime start, LocalDateTime end) {
        long daysBetween = Duration.between(start, end).toDays();

        if(daysBetween > 365) {
            throw new ResponseStatusException(HttpStatus.UNPROCESSABLE_CONTENT, "Timestamp cannot be bigger than 365 days");
        }

    }

    public boolean isStringNull(String str) {
        if(str == null || str.isEmpty()) return true;
        return false;
    }
}

File: front/src/services/AccountHistoryService.ts
import type { AccountHistory } from "@/models/AccountHistory";
import { accountHistory } from "./seed";

const LS_KEY = 'history';

export type TransactionType = 'deposito' | 'saque' | 'transferencia';

export interface AccountHistoryDTO {
    accountNumber: string;
    type: TransactionType;
    originClientCpf: string;
    originClientName: string;
    destinationClientCpf: string | null;
    destinationClientName: string | null;
    amount: number;
    managerCpf?: string | null;
    managerName?: string | null;
}

export interface ExtractDTO {
    saldoAbertura: string;
    movimentacoes: AccountHistory[];
}

export class AccountHistoryService {

    listAll(): AccountHistory[] {
        const history = localStorage[LS_KEY];
        if (!history) {
            localStorage.setItem(LS_KEY, JSON.stringify(accountHistory))
            return [...accountHistory]
        }
        return history ? JSON.parse(history) : [];
    }

    private persist(history: AccountHistory[]): void {
        localStorage[LS_KEY] = JSON.stringify(history);
    }

    insert(accountHistory: AccountHistory): void {
        const history = this.listAll();
        accountHistory.id = new Date().getTime();
        history.push(accountHistory);
        this.persist(history);
    }

    findById(id: number): AccountHistory | undefined {
        return this.listAll().find(h => h.id === id);
    }

    findByClient(clientCpf: string): AccountHistory[] {
        return this.listAll().filter(h => h.originClientCpf === clientCpf);
    }

    findByAccountNumber(accountNumber: string): AccountHistory[] {
        return this.listAll().filter(h => h.accountNumber === accountNumber);
    }

    createAcountHistory(dto: AccountHistoryDTO): AccountHistory {
        const accountHistory: AccountHistory = {
            id: 0,
            accountNumber: dto.accountNumber,
            type: dto.type,
            originClientCpf: dto.originClientCpf,
            originClientName: dto.originClientName,
            destinationClientCpf: dto.destinationClientCpf,
            destinationClientName: dto.destinationClientName,
            amount: dto.amount,
            createdAt: new Date(),
        } as unknown as AccountHistory;

        this.insert(accountHistory);
        return accountHistory;
    }

    private parseDateTime(dateStr: string | null | undefined, isStart: boolean): Date {
        if (this.isStringNull(dateStr)) {
            const d = new Date();
            if (isStart) {
                d.setDate(d.getDate() - 30);
                d.setHours(0, 0, 0, 0);
            } else {
                d.setHours(23, 59, 59, 999);
            }
            return d;
        }

        const trimmed = dateStr!.trim();

        if (trimmed.length === 10) {
            const d = new Date(`${trimmed}T00:00:00`);
            if (!isStart) d.setHours(23, 59, 59, 999);
            return d;
        }

        if (trimmed.includes(' ') && !trimmed.includes('T')) {
            return new Date(trimmed.replace(' ', 'T'));
        }

        return new Date(trimmed);
    }

    getExtract(
        accountNumber: string,
        start: string | null,
        end: string | null,
        userCPF: string
    ): ExtractDTO {
        const startDate = this.parseDateTime(start, true);
        const endDate = this.parseDateTime(end, false);

        this.checkDate(startDate, endDate);

        const saldoAbertura = this.getInitialBalance(accountNumber, startDate, userCPF);

        const movimentacoes = this.listAll()
            .filter(h =>
                h.accountNumber === accountNumber &&
                new Date(h.createdAt) >= startDate &&
                new Date(h.createdAt) <= endDate
            );

        return {
            saldoAbertura: saldoAbertura.toFixed(2),
            movimentacoes,
        };
    }

    getInitialBalance(accountNumber: string, start: Date, userCPF: string): number {
        const history = this.listAll()
            .filter(h =>
                h.accountNumber === accountNumber &&
                new Date(h.createdAt) < start
            );

        let balance = 0;

        for (const ac of history) {
            const value = Number(ac.amount);

            if (ac.type === 'deposito') {
                balance += value;
            } else if (ac.type === 'saque') {
                balance -= value;
            } else if (ac.type === 'transferencia') {
                if (ac.destinationClientCpf === userCPF) balance += value;
                if (ac.originClientCpf === userCPF) balance -= value;
            }
        }

        return balance;
    }

    checkDate(start: Date, end: Date): void {
        const daysBetween = (end.getTime() - start.getTime()) / (1000 * 60 * 60 * 24);

        if (daysBetween > 365) {
            throw new Error('Filtro não pode ter diferença maior que 365 dias');
        }
    }

    isStringNull(str: string | null | undefined): boolean {
        return str == null || str.length === 0;
    }
}


